panic: concurrent write to websocket connection

```
panic: concurrent write to websocket connection

goroutine 1963 [running]:
github.com/gorilla/websocket.(*messageWriter).flushFrame(0xc000052ee8, 0x1, {0xc0001a7bf3?, 0xc0001a7bc0?, 0x33?})
	/home/chen/go/pkg/mod/github.com/gorilla/websocket@v1.5.3/conn.go:617 +0x4b8
github.com/gorilla/websocket.(*Conn).WriteMessage(0xc00101c160, 0xc0001a7b80?, {0xc0001a7bc0, 0x33, 0x40})
	/home/chen/go/pkg/mod/github.com/gorilla/websocket@v1.5.3/conn.go:770 +0x126
main.broadcast_ws_msg.func1(0xc00101c160)
	/run/test/cloud-clip.go/server-go/cloud-clip/ws.go:31 +0x94
created by main.broadcast_ws_msg in goroutine 1960
	/run/test/cloud-clip.go/server-go/cloud-clip/ws.go:29 +0x293
```

<br>

`websocket.(*Conn).WriteMessage` is not thread-safe

https://github.com/gorilla/websocket/issues/119

--- a/server-go/cloud-clip/main.go
+++ b/server-go/cloud-clip/main.go
@@ -15,14 +15,23 @@ import (
 	"strings"
 	"time"
 
-	"github.com/gorilla/websocket"
+	websocket "github.com/yurenchen000/cloud-clipboard/server-go/cloud-clip/safews"
+	// "github.com/gorilla/websocket"
 	"github.com/spaolacci/murmur3"
 )
 
 var (
-	upgrader = websocket.Upgrader{
-		CheckOrigin: func(r *http.Request) bool {
-			return true
+	// upgrader = websocket.Upgrader{
+	// 	CheckOrigin: func(r *http.Request) bool {
+	// 		return true
+	// 	},
+	// }
+
+	upgrader = websocket.Upgrader{ // inherit from orig websocket.Upgrader
+		Upgrader: &websocket.OrigUpgrader{
+			CheckOrigin: func(r *http.Request) bool {
+				return true
+			},
 		},
 	}
 

--- a/server-go/cloud-clip/safews/safe_ws.go
+++ b/server-go/cloud-clip/safews/safe_ws.go
@@ -1,0 +1,96 @@
+// a gorilla/websocket wrapper for safe concurrent write
+package SafeWebsocket
+
+import (
+	"errors"
+	"log"
+
+	"net/http"
+
+	"github.com/gorilla/websocket"
+)
+
+// --------------- ws var
+var TextMessage = websocket.TextMessage
+
+// --------------- ws upgrader
+type OrigUpgrader = websocket.Upgrader
+
+// Upgrader 继承 websocket.Upgrader
+type Upgrader struct {
+	*websocket.Upgrader // 嵌入原始 Upgrader
+}
+
+// Upgrade 重写方法，返回安全的 SafeWebSocket 连接
+func (u *Upgrader) Upgrade(w http.ResponseWriter, r *http.Request, responseHeader http.Header) (*Conn, error) {
+	// 调用原始 Upgrade 获取标准 websocket 连接
+	rawConn, err := u.Upgrader.Upgrade(w, r, responseHeader)
+	if err != nil {
+		return nil, err
+	}
+
+	// 封装为线程安全的 SafeWebSocket
+	return NewConn(rawConn), nil
+}
+
+// --------------- ws conn
+
+type Conn struct {
+	// conn      *websocket.Conn
+	*websocket.Conn
+	writeChan chan writeRequest //
+	closeChan chan struct{}     //only close: boradcast writeChan closed
+}
+
+type writeRequest struct {
+	messageType int
+	data        []byte
+}
+
+func NewConn(conn *websocket.Conn) *Conn {
+	ws := &Conn{
+		Conn:      conn,
+		writeChan: make(chan writeRequest, 100),
+		closeChan: make(chan struct{}),
+	}
+	go ws.writePump()
+	return ws
+}
+
+func (ws *Conn) writePump() {
+	for req := range ws.writeChan {
+		err := ws.Conn.WriteMessage(req.messageType, req.data)
+		if err != nil {
+			log.Println("write error:", err)
+			ws.Close()
+			// delete(ws_list, ws)
+		}
+	}
+}
+
+var errChanClosed = errors.New("sws: already closed")
+
+func (ws *Conn) WriteMessage(messageType int, data []byte) error {
+	select {
+	case <-ws.closeChan: //closed: do nothing
+		return errChanClosed
+	case ws.writeChan <- writeRequest{messageType, data}: //closed: panic
+		return nil
+	}
+}
+func (ws *Conn) Close() error {
+	select {
+	case <-ws.closeChan: //closed: do nothing
+		return errChanClosed
+	default:
+		close(ws.closeChan)
+		writeChan := ws.writeChan
+		ws.writeChan = nil //avoid write to closed
+		close(writeChan)   //avoid for range leak
+		return ws.Conn.Close()
+		// return nil
+	}
+}
+
+// chan 回收机制: 不是关闭后被回收, 而是不再被引用 被回收
+// 可以显式 ch = nil, 也可以等待 保存 ch 的结构被释放

--- a/server-go/cloud-clip/ws.go
+++ b/server-go/cloud-clip/ws.go
@@ -13,7 +13,8 @@ import (
 	"strings"
 	"sync"
 
-	"github.com/gorilla/websocket"
+	websocket "github.com/yurenchen000/cloud-clipboard/server-go/cloud-clip/safews"
+	// "github.com/gorilla/websocket"
 	"github.com/ua-parser/uap-go/uaparser"
 )
 

